teps/0142-enable-step-reusability.md

```diff
@@ -0,0 +1,997 @@
+    - [Prior Art](#prior-art)
+      - [1.TaskGroup Custom Task](#1-taskgroup-custom-task)
+      - [2.Reuse Steps](#2-reuse-steps)
+      - [3.Pre/Post Steps](#3-prepost-steps)
+      - [4.Pipeline in a Pod](#4-pipeline-in-a-pod)
+    - [CI/CD Systems based on Tekton](#cicd-systems-based-on-tekton)
+      - [1.Nubank Workflows](#1-nubank-workflows)
+      - [2.JenkinsX](#2-jenkinsx) 
+    - [Comparable CI/CD Systems](#comparable-cicd-systems)
+      - [1.Github Actions](#1-github-actions)
+- [Proposal](#proposal) 
+  - [StepAction CRD](#stepfoo-crd)
+  - [Referenced Steps](#referenced-steps)
+    - [Params](#params)
+    - [Results](#results)
+    - [Workspaces](#workspaces)
+  - [Inlined Steps](#inlined-steps)
+- [Secure Software Supply Chain](#secure-software-supply-chain)
+  - [Trusted Steps](#trusted-steps)
+  - [Provenance](#provenance)
+- [Possible Names](#possible-names)
+- [Future Work](#future-work)
+  - [Sidecar](#sidecar)
+- [Alternatives](#alternatives)
+  - [StepCRD](#stepcrd)
+  - [ActionCRD](#actioncrd)
+  - [Tasks in Tasks](#tasks-in-tasks)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes to introduce a `StepAction` CRD as the smallest, scriptable and reusable unit of work in Tekton. 
+
+The reusable units of work in Tekton, `Tasks`, currently cannot natively execute in the same environment with a shared file system. As such, Tekton users have to choose between reusability and performance. This design doc revisits the components and architecture of Tekton to both enable reusability and optimize performance. 
+
+### Background
+
+`Tasks` are the reusable unit of work in Tekton. A `Task` is made up of a sequence of `Steps`. A `Task` and its `Steps` execute in Kubernetes as a `Pod` made up of a sequence of `Containers`. The `Steps` of a `Task` have access to shared storage and resources, such as `Parameters` and `Workspaces`. `Tasks` are combined in `Pipelines` which are graphs where each node represents a `Task`. The `Tasks` in a `Pipeline` execute in separate `Pods`, thus they need to share data via a `Workspace` that is generally backed by a `Persistent Volume`.
+
+![Components](images/0142-components.png)
+
+## Motivation
+
+`Steps` are the smallest unit of work but they aren’t reusable because they are specified in `Tasks`. `Task` authors end up writing a lot of general purpose `Tasks` with single `Step` to make them reusable for most users. In fact, about 77% of `Tasks` in the Tekton Catalog have a single `Step` only. However, these reusable units cannot execute in a shared context – `Pod`. 
+
+When users need to combine `Steps` to execute together, they are forced to choose between reusability and performance. If they prioritize performance over reusability, they would copy and paste the `Steps` of the `Tasks` into a new `Task` to execute in one `Pod` with a shared file system. If they prioritize reusability over performance, they would execute the `Tasks` with single `Steps` in separate `Pods` without a shared file system.
```

Comment:
It's true for most of the cases but there exists a significant number of use cases where performance with step CRD would degrade.  These use cases cater to running Taskrun in parallel.

Reply:
Absolutely! I don't think here, were are suggesting one or the other. A Step is incomplete without a Task since it cannot be executed on its own. It's still a part of the TaskRun so the user has full capability to execute a Task with a single step. It's just that this Step does not have to be inlined. It would be nice if it could be reused.

Reply:
Running one huge `pod` with a series of containers running in sequence might not gain as much performance as expected.

For a common CI/CD use case, a simple application repo, git-clone can be the first step followed by building an image from that repo.

And a pipeline building many jar files as part of a single build process might be expensive and not feasible for a single pod.

Reply:
Agreed that there are many cases where running sequentially degrades performance. 

The point here is that the moment a user wants to combine steps that they reuse, they have to choose between `copy-and-pasting` (the usability factor) or write them as tasks (to get that reusability) and put those tasks in sequence (which hinders performance compared to two steps in sequence). This statement is targeting simple use-cases where users are really forced to write a pipeline where it makes more sense to simply write a Task.

Reply:
NIT: The "When users need to combine `Steps` to execute together" part clarifies that this is not for all use cases, but maybe it could be made a bit more explicit. 

There are many different use cases, depending on this shape, duration and data I/O of each workload it makes more sense to spread them around in different pods and pay the pod setup cost and data I/O to share data, or keep them together in a pod, and pay the reusability, duration and scalability cost that goes with it. This proposal helps reducing the "reusability" cost from the second option, which I think is a great advancement for Tekton.